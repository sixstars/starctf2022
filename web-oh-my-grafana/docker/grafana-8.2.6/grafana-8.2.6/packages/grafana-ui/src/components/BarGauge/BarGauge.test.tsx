import React from 'react';
import { shallow } from 'enzyme';
import {
  DisplayValue,
  VizOrientation,
  ThresholdsMode,
  FALLBACK_COLOR,
  Field,
  FieldType,
  getDisplayProcessor,
  createTheme,
} from '@grafana/data';
import {
  BarGauge,
  Props,
  getCellColor,
  getValueColor,
  getBasicAndGradientStyles,
  getBarGradient,
  getTitleStyles,
  getValuePercent,
  BarGaugeDisplayMode,
  calculateBarAndValueDimensions,
} from './BarGauge';

const green = '#73BF69';
const orange = '#FF9830';

function getProps(propOverrides?: Partial<Props>): Props {
  const field: Partial<Field> = {
    type: FieldType.number,
    config: {
      min: 0,
      max: 100,
      thresholds: {
        mode: ThresholdsMode.Absolute,
        steps: [
          { value: -Infinity, color: 'green' },
          { value: 70, color: 'orange' },
          { value: 90, color: 'red' },
        ],
      },
    },
  };
  const theme = createTheme();
  field.display = getDisplayProcessor({ field, theme });

  const props: Props = {
    displayMode: BarGaugeDisplayMode.Basic,
    field: field.config!,
    display: field.display!,
    height: 300,
    width: 300,
    value: field.display(25),
    theme,
    orientation: VizOrientation.Horizontal,
  };

  Object.assign(props, propOverrides);
  return props;
}

const setup = (propOverrides?: object) => {
  const props = getProps(propOverrides);
  const wrapper = shallow(<BarGauge {...props} />);
  const instance = wrapper.instance() as BarGauge;

  return {
    instance,
    wrapper,
  };
};

function getValue(value: number, title?: string): DisplayValue {
  return { numeric: value, text: value.toString(), title: title };
}

describe('BarGauge', () => {
  describe('getCellColor', () => {
    it('returns a fallback if the positionValue is null', () => {
      const props = getProps();
      expect(getCellColor(null, props.value, props.display)).toEqual({
        background: FALLBACK_COLOR,
        border: FALLBACK_COLOR,
      });
    });

    it('does not show as lit if the value is null (somehow)', () => {
      const props = getProps();
      expect(getCellColor(1, (null as unknown) as DisplayValue, props.display)).toEqual(
        expect.objectContaining({
          isLit: false,
        })
      );
    });

    it('does not show as lit if the numeric value is NaN', () => {
      const props = getProps();
      expect(
        getCellColor(
          1,
          {
            numeric: NaN,
            text: '0',
          },
          props.display
        )
      ).toEqual(
        expect.objectContaining({
          isLit: false,
        })
      );
    });

    it('does not show as lit if the positionValue is greater than the numeric value', () => {
      const props = getProps();
      expect(getCellColor(75, props.value, props.display)).toEqual(
        expect.objectContaining({
          isLit: false,
        })
      );
    });

    it('shows as lit otherwise', () => {
      const props = getProps();
      expect(getCellColor(1, props.value, props.display)).toEqual(
        expect.objectContaining({
          isLit: true,
        })
      );
    });

    it('returns a fallback if there is no display processor', () => {
      const props = getProps();
      expect(getCellColor(null, props.value, undefined)).toEqual({
        background: FALLBACK_COLOR,
        border: FALLBACK_COLOR,
      });
    });
  });

  describe('Get value color', () => {
    it('should get the threshold color if value is same as a threshold', () => {
      const props = getProps();
      props.value = props.display!(70);
      expect(getValueColor(props)).toEqual(orange);
    });
    it('should get the base threshold', () => {
      const props = getProps();
      props.value = props.display!(-10);
      expect(getValueColor(props)).toEqual(green);
    });
  });

  describe('Get value percent', () => {
    it('0 to 100 and value 40', () => {
      expect(getValuePercent(40, 0, 100)).toEqual(0.4);
    });

    it('50 to 100 and value 75', () => {
      expect(getValuePercent(75, 50, 100)).toEqual(0.5);
    });

    it('-30 to 30 and value 0', () => {
      expect(getValuePercent(0, -30, 30)).toEqual(0.5);
    });

    it('-30 to 30 and value 30', () => {
      expect(getValuePercent(30, -30, 30)).toEqual(1);
    });
  });

  describe('Vertical bar without title', () => {
    it('should not include title height in height', () => {
      const props = getProps({
        height: 300,
        value: getValue(100),
        orientation: VizOrientation.Vertical,
      });
      const styles = getBasicAndGradientStyles(props);
      expect(styles.bar.height).toBe('270px');
    });
  });

  describe('Vertical bar with title', () => {
    it('should include title height in height', () => {
      const props = getProps({
        height: 300,
        value: getValue(100, 'ServerA'),
        orientation: VizOrientation.Vertical,
      });
      const styles = getBasicAndGradientStyles(props);
      expect(styles.bar.height).toBe('249px');
      expect(styles.emptyBar.bottom).toBe('-3px');
    });
  });

  describe('Horizontal bar', () => {
    it('should stretch items', () => {
      const props = getProps({
        height: 300,
        value: getValue(100, 'ServerA'),
        orientation: VizOrientation.Horizontal,
      });
      const styles = getBasicAndGradientStyles(props);
      expect(styles.wrapper.alignItems).toBe('stretch');
      expect(styles.emptyBar.left).toBe('-3px');
    });
  });

  describe('Horizontal bar with title', () => {
    it('should place above if height > 40', () => {
      const props = getProps({
        height: 41,
        value: getValue(100, 'AA'),
        orientation: VizOrientation.Horizontal,
      });
      const styles = getTitleStyles(props);
      expect(styles.wrapper.flexDirection).toBe('column');
    });
  });

  describe('Horizontal bar with title', () => {
    it('should place below if height < 40', () => {
      const props = getProps({
        height: 30,
        value: getValue(100, 'AA'),
        orientation: VizOrientation.Horizontal,
      });
      const styles = getTitleStyles(props);
      expect(styles.wrapper.flexDirection).toBe('row');
    });

    it('should calculate title width based on title', () => {
      const props = getProps({
        height: 30,
        value: getValue(100, 'AA'),
        orientation: VizOrientation.Horizontal,
      });
      const styles = getTitleStyles(props);
      expect(styles.title.width).toBe('17px');

      const props2 = getProps({
        height: 30,
        value: getValue(120, 'Longer title with many words'),
        orientation: VizOrientation.Horizontal,
      });
      const styles2 = getTitleStyles(props2);
      expect(styles2.title.width).toBe('43px');
    });

    it('should use alignmentFactors if provided', () => {
      const props = getProps({
        height: 30,
        value: getValue(100, 'AA'),
        alignmentFactors: {
          title: 'Super duper long title',
          text: '1000',
        },
        orientation: VizOrientation.Horizontal,
      });
      const styles = getTitleStyles(props);
      expect(styles.title.width).toBe('37px');
    });
  });

  describe('Gradient', () => {
    it('should build gradient based on thresholds', () => {
      const props = getProps({ orientation: VizOrientation.Vertical, value: getValue(100) });
      const gradient = getBarGradient(props, 300);
      expect(gradient).toBe('linear-gradient(0deg, #73BF69, #73BF69 105px, #FF9830 240px, #F2495C)');
    });

    it('should stop gradient if value < threshold', () => {
      const props = getProps({ orientation: VizOrientation.Vertical, value: getValue(70) });
      const gradient = getBarGradient(props, 300);
      expect(gradient).toBe('linear-gradient(0deg, #73BF69, #73BF69 105px, #FF9830)');
    });
  });

  describe('Render with basic options', () => {
    it('should render', () => {
      const { wrapper } = setup();
      expect(wrapper).toMatchSnapshot();
    });
  });

  describe('calculateBarAndValueDimensions', () => {
    it('valueWidth should including paddings in valueWidth', () => {
      const result = calculateBarAndValueDimensions(
        getProps({
          height: 30,
          width: 100,
          value: getValue(1, 'AA'),
          orientation: VizOrientation.Horizontal,
        })
      );
      expect(result.valueWidth).toBe(21);
    });
  });
});
