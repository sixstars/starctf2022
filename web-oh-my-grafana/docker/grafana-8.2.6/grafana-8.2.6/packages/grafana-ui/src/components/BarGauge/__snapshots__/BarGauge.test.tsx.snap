// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarGauge Render with basic options should render 1`] = `
<div
  style={
    Object {
      "display": "flex",
      "flexDirection": "column",
      "overflow": "hidden",
      "width": "100%",
    }
  }
>
  <div
    style={
      Object {
        "alignItems": "stretch",
        "display": "flex",
        "flexDirection": "row-reverse",
        "flexGrow": 1,
        "justifyContent": "flex-end",
      }
    }
  >
    <FormattedDisplayValue
      aria-label="Bar gauge value"
      style={
        Object {
          "alignItems": "center",
          "color": "#73BF69",
          "display": "flex",
          "fontSize": 140,
          "height": "300px",
          "justifyContent": "flex-end",
          "lineHeight": 1,
          "paddingLeft": "10px",
          "paddingRight": "10px",
          "width": "60px",
        }
      }
      value={
        Object {
          "color": "#73BF69",
          "numeric": 25,
          "percent": 0.25,
          "prefix": undefined,
          "suffix": undefined,
          "text": "25",
        }
      }
    />
    <div
      style={
        Object {
          "background": "rgba(255,255,255, 0.07)",
          "borderRadius": "3px",
          "display": "flex",
          "flexGrow": 1,
          "left": "-3px",
          "position": "relative",
        }
      }
    />
    <div
      style={
        Object {
          "background": "rgba(115, 191, 105, 0.35)",
          "borderRadius": "3px",
          "borderRight": "2px solid #73BF69",
          "height": "300px",
          "position": "relative",
          "transition": "width 1s",
          "width": "60px",
          "zIndex": 1,
        }
      }
    />
  </div>
</div>
`;
